src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -70,20 +71,157 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
+      instance.componentWillMount();
+      // If we had additional state updates during this life-cycle, let's
+      // process them now.
+      const updateQueue = workInProgress.updateQueue;
+      if (updateQueue) {
+        instance.state = mergeUpdateQueue(updateQueue, state, props);
+      }
+    }
+  }
+
+  // Called on a preexisting class instance. Returns false if a resumed render
+  // could be reused.
+  function resumeMountClassInstance(workInProgress : Fiber) : boolean {
+    const instance = workInProgress.stateNode;
+    const newState = workInProgress.memoizedState;
+    let newProps = workInProgress.pendingProps;
+    if (!newProps) {
+      // If there isn't any new props, then we'll reuse the memoized props.
+      // This could be from already completed work.
+      newProps = workInProgress.memoizedProps;
+      if (!newProps) {
+        throw new Error('There should always be pending or memoized props.');
+      }
+    }
+
+    // TODO: Should we deal with a setState that happened after the last
+    // componentWillMount and before this componentWillMount? Probably
+    // unsupported anyway.
+
+    const updateQueue = workInProgress.updateQueue;
+
+    // If this completed, we might be able to just reuse this instance.
+    if (typeof instance.shouldComponentUpdate === 'function' &&
+        !(updateQueue && updateQueue.isForced) &&
+        workInProgress.memoizedProps !== null &&
+        !instance.shouldComponentUpdate(newProps, newState)) {
+      return false;
+    }
+
+    // If we didn't bail out we need to construct a new instance. We don't
+    // want to reuse one that failed to fully mount.
+    const newInstance = constructClassInstance(workInProgress);
+    if (typeof newInstance.componentWillMount === 'function') {
+      newInstance.componentWillMount();
+      // If we had additional state updates during this life-cycle, let's
+      // process them now.
+      const newUpdateQueue = workInProgress.updateQueue;
+      if (newUpdateQueue) {
+        instance.state = mergeUpdateQueue(newUpdateQueue, newState, newProps);
+      }
+    }
+    return true;
+  }
+
+  // Invokes the update life-cycles and returns false if it shouldn't rerender.
+  function updateClassInstance(current : Fiber, workInProgress : Fiber) : boolean {
+    const instance = workInProgress.stateNode;
+
+    const oldProps = current.memoizedProps;
+    let newProps = workInProgress.pendingProps;
+    if (!newProps) {
+      // If there aren't any new props, then we'll reuse the memoized props.
+      // This could be from already completed work.
+      newProps = workInProgress.memoizedProps;
+      if (!newProps) {
+        throw new Error('There should always be pending or memoized props.');
+      }
+    }
+
+    // Note: During these life-cycles, instance.props/instance.state are what
+    // ever the previously attempted to render - not the "current". However,
+    // during componentDidUpdate we pass the "current" props.
```

Comment:
Like discussed in chat, this is intended. It's a bit confusing but it's the most likely semantics that will work with current code.

The primary goal is to have a nicer upgrade path from existing code. Ease of understanding the semantics of the life-cycles can be achieved by redesigning a new API for them.